src/Analyzer/Passes/IfTransformStringsToEnumPass.cpp

```diff
@@ -0,0 +1,306 @@
+
+    {
+        auto enum_literal = std::make_shared<ConstantValue>(fmt::format("Array({})", enum_string), std::make_shared<DataTypeString>());
+        auto enum_literal_node = std::make_shared<ConstantNode>(std::move(enum_literal));
+
+        auto cast_function = FunctionFactory::instance().get("_CAST", context);
+        QueryTreeNodes arguments{array_to, enum_literal_node};
+
+        auto function_node = std::make_shared<FunctionNode>("_CAST");
+        function_node->resolveAsFunction(std::move(cast_function), std::make_shared<DataTypeArray>(result_type));
+        function_node->getArguments().getNodes() = std::move(arguments);
+
+        array_to = std::move(function_node);
+    }
+
+    {
+        auto enum_literal = std::make_shared<ConstantValue>(enum_string, std::make_shared<DataTypeString>());
+        auto enum_literal_node = std::make_shared<ConstantNode>(std::move(enum_literal));
+
+        auto cast_function = FunctionFactory::instance().get("_CAST", context);
+        QueryTreeNodes arguments{default_value, enum_literal_node};
+
+        auto function_node = std::make_shared<FunctionNode>("_CAST");
+        function_node->resolveAsFunction(std::move(cast_function), result_type);
+        function_node->getArguments().getNodes() = std::move(arguments);
+
+        default_value = std::move(function_node);
+    }
+}
+
+class ConvertStringsToEnumVisitor : public InDepthQueryTreeVisitor<ConvertStringsToEnumVisitor>
+{
+public:
+    explicit ConvertStringsToEnumVisitor(std::unordered_set<std::string_view> used_functions_, ContextPtr context_)
+        : used_functions(std::move(used_functions_)), context(std::move(context_))
+    {
+    }
+
+    static bool needChildVisit(VisitQueryTreeNodeType & parent, VisitQueryTreeNodeType & /* child */)
+    {
+        return parent->getNodeType() != QueryTreeNodeType::FUNCTION;
+    }
+
+    void visitImpl(QueryTreeNodePtr & node)
+    {
+        auto * function_node = node->as<FunctionNode>();
+
+        if (!function_node)
+            return;
+
+        /// we cannot change the type of its result because it's used
+        /// as argument in another function
+        if (used_functions.contains(function_node->getAlias()))
+            return;
+
+        std::string_view function_name = function_node->getFunctionName();
+        if (function_name == "if")
+        {
+            auto & argument_nodes = function_node->getArguments().getNodes();
+
+            if (argument_nodes.size() != 3)
+                return;
+
+            const auto * first_literal = argument_nodes[1]->as<ConstantNode>();
+            const auto * second_literal = argument_nodes[2]->as<ConstantNode>();
+
+            if (!first_literal || !second_literal)
+                return;
+
+            if (!WhichDataType(first_literal->getResultType()).isString() || !WhichDataType(second_literal->getResultType()).isString())
+                return;
+
+            std::set<std::string> string_values;
```

Comment:
Consider to make this a member that will be cleared during visitImpl.

Reply:
What would be the reason for doing that?